pkg/controller/node/nodecontroller.go

```diff
@@ -153,8 +167,12 @@ type NodeController struct {
 	// workers that evicts pods from unresponsive nodes.
 	zonePodEvictor map[string]*RateLimitedTimedQueue
 	// workers that are responsible for tainting nodes.
-	zoneNotReadyOrUnreachableTainer map[string]*RateLimitedTimedQueue
-	podEvictionTimeout              time.Duration
+	zoneNoExecuteTainer map[string]*RateLimitedTimedQueue
+
+	noScheduleTainerLock sync.Mutex
+	zoneNoScheduleTainer map[string]*RateLimitedTimedQueue
```

Comment:
I'll need to read it carefully (I'll try to do this in the evening, but I may fail, so I'm leaving this note here), but I think this will be wrong. We want to have evictor (i.e. rate-limited tainting) only for Unreachable/NotReady Taints. The rest of the taints should not need rate limiting as they're NoSchedule ones. The only thing that we should protect against is TaintFlapping (i.e. adding and removing Taints in quick succession).

As I said - it's possible that this is how it's done, but I didn't have time to read it carefully enough this time.